### ISSUE TITLE
Python: change generated file layout to all easier re-generation

### ISSUE THREAD

**Issue by zaibon** [Jumpscale/go-raml#393]
Currently it is really hard to re-generate the code once it has been generated once.
We need to find a way to clearly show the difference between the generated files, likes types, that should not be edited and the files that requires the user to modify them.

So the idea would be to have something like:
```
├── api_spec
│   ├── main.raml
│   ├── resourcetypes.raml
│   ├── traits.raml
│   └── types.raml
├── generated
│   ├── client
│   └── server
├── server
│   ├── app.py
│   ├── group_api.py
│   └── user_api.py
```

`api_spec` directory contains the raml files.
`generated` contains 2 directories, 1 for server 1 for client
`server` contains only the files the user need to modify.

When updating the raml, and generating again. only the files in generated directory will be touch. So you don't risk any trouble. This allow to re-generated the code as much as we want without risk.

There is still one problem with this approach though. It works well if you change the types, but it still doesn't solve the case when you add/remove a method. Cause then the files in server directory needs to be impacted. Something we need to thing solve still.

This works well for python. We need to check if a similar approach can be done for Go and lua.

Attached to this issue an example of how the code should look like
[goraml.zip](https://github.com/Jumpscale/go-raml/files/1590903/goraml.zip)

**Comment by serboctor** [comment]
I still have problems with this approach:
- The types will actually not be regenerated unless if we change the override flag in the generate function. We need to revisit all the calls of the generate functions and decide which files need to be overridden otherwise this change wont fix anything, it will merely be a different structure for the directories.
- raml doesnt support all types we might want to generate, for example it doesn't have uint. As a result sometimes we modify types after they're generated. Do we regenerate and override the files and make it the user responsibility to edit them again? or do we depend on adding invalid types in the raml spec?
-  If we initially generate the <resource>_api.py files in the ./server dir and we dont regenerate when a file exists, the user won't really know that some funcitons were added. However, if we generate it in ./generated/server/ and move it to ./server, at least the user will realize there are new functions when he sees that a file hasn't been copied and can manually copy the methods. Or may be we need to think of a third solution. But the way I see it, this current approach still has a lot of problems.

**Comment by zaibon** [comment]
> raml doesnt support all types we might want to generate, for example it doesn't have uint.

This is tricky cause supporting non valid raml type would prevent us to use any other tool on our raml files. So I guess letting the user the responsibility to make it right again the way to go as long as we don't have a better solution

> However, if we generate it in ./generated/server/ and move it to ./server, at least the user will realize there are new functions when he sees that a file hasn't been copied and can manually copy the methods. 

Yep that's better cause then he can use any diff toll and clearly see what's new. Let's go for that.

**Comment by serboctor** [comment]
Ok so to sum up here is what will be done for the python generation:

1- All types/object classes will be regenerated even if they exist under ./generated/server. It will be the user's responsibility to re-add any changes that were previously manually added
2- api files will be generated under ./generated/server then copied to ./server IF they don't exist there. It will be the user's responsibility to check if there are new functions and to clean up the extra files under ./generated/server
3- Add clear comments to all files stating whether this file will be overwritten when regenerating or if the file can be safely edited by the user.

**Comment by iwanbk** [comment]
> 1- All types/object classes will be regenerated even if they exist under ./generated/server. It will be the user's responsibility to re-add any changes that were previously manually added

agree

> 2- api files will be generated under ./generated/server then copied to ./server IF they don't exist there. It will be the user's responsibility to check if there are new functions and to clean up the extra files under ./generated/server

i'm thinking that user shouldn't need to modify the generated API files.

The generated API files will look like this
```python
import handler.getGroupHandler
import handler.updateGroupHandler

@group_api.route('/group/<id>', methods=['GET'])
def getGroup(id):
    '''
    Get group, id=int
    It is handler for GET /group/<id>
    '''

    return getGroupHandler.Handle(id)


@group_api.route('/group/<id>', methods=['POST'])
def updateGroup(id):
    '''
    Update group
    It is handler for POST /group/<id>
    '''

    inputs = request.get_json()
    try:
        Group_schema_validator.validate(inputs)
    except jsonschema.ValidationError as e:
        return jsonify(errors="bad request body"), 400

    return updateGroupHandler.handle(inputs, id)
```

The API files will always be regenerated, while the handler files only generated if not exist.
With this way, we can minimize the number of LoC to check when we regenerate the files.

**Comment by iwanbk** [comment]
> This works well for python. We need to check if a similar approach can be done for Go and lua.

I think we should create separate issue for each language so we could split the work.
But other langs should be blocked by this issue because i'm sure other langs could learn from the python solution

**Comment by serboctor** [comment]
> The API files will always be regenerated, while the handler files only generated if not exist.
With this way, we can minimize the number of LoC to check when we regenerate the files.

This is similar to how go-swagger does things which I already like. If this will solve all our regenerations issue and we won't need to copy stuff around, should we go back to the old dir structure? I already don't love the ./generated and ./server structure.

**Comment by iwanbk** [comment]
> If this will solve all our regenerations issue and we won't need to copy stuff around, should we go back to the old dir structure? I already don't love the ./generated and ./server structure.

Yep, i vote to back to old dir structure

**Comment by zaibon** [comment]
I like the idea from iwan.
Only problem I have with it is that user can't choose how the BadRequest error is send back to the client
maybe this part needs to be part of the handler too ?

**Comment by iwanbk** [comment]
> Only problem I have with it is that user can't choose how the BadRequest error is send back to the client


Not sure i understand your concern, we already have this in the generated code
```
except jsonschema.ValidationError as e:
        return jsonify(errors="bad request body"), 400
```

**Comment by serboctor** [comment]
> Not sure i understand your concern, we already have this in the generated code

But sometimes we edit this part in the generated code, so may be we can generate this part in the handler instead so the user is free to change it or not.

**Comment by iwanbk** [comment]
> But sometimes we edit this part in the generated code

Why **edit** this part?

I can imagine that user might need to **add** more validation, though, which should be added to the handler

**Comment by zaibon** [comment]
cause most of the time I don't want to handle my error with 
`jsonify(errors="bad request body"), 400`

you could have some completly different error type returned or even do something different then jsut a 400
